Struct with `fn()` field causes `unsupported type` error

The following code:
```Rust
pub struct WierdStruct {
    f: fn(),
}
```
crashes flux with the error:
```
error[E0999]: refinement of unsupported struct
   --> src/main.rs:229:12
    |
229 | pub struct WierdStruct {
    |            ^^^^^^^^^^^ this struct contains unsupported features
    |
    = note: unsupported type: `fn()`
```

This pattern can be found in Tock's `Iter<...>` struct ([link](https://github.com/PLSysSec/tock/blob/9ff6edd93a5ec679ab4650adb45c250ba7d95d62/kernel/src/grant.rs#L1792-L1808)) which uses it inside a `core::iter::FilterMap` field.

--- a/crates/flux-desugar/src/desugar.rs
+++ b/crates/flux-desugar/src/desugar.rs
@@ -568,7 +568,7 @@ impl<'a, 'genv, 'tcx: 'genv> RustItemCtxt<'a, 'genv, 'tcx> {
             };
             // Fix up the span in asyncness
             if let surface::Async::Yes { span, .. } = fn_sig.asyncness {
-                header.asyncness = hir::IsAsync::Async(span)
+                header.asyncness = hir::IsAsync::Async(span);
             }
             (generics, decl)
         } else {

--- a/crates/flux-desugar/src/desugar.rs
+++ b/crates/flux-desugar/src/desugar.rs
@@ -16,8 +16,7 @@ use hir::{def::DefKind, ItemKind};
 use rustc_data_structures::{fx::FxIndexSet, unord::UnordMap};
 use rustc_errors::{Diagnostic, ErrorGuaranteed};
 use rustc_hash::{FxHashMap, FxHashSet};
-use rustc_hir as hir;
-use rustc_hir::OwnerId;
+use rustc_hir::{self as hir, OwnerId, ParamName};
 use rustc_span::{
     def_id::{DefId, LocalDefId},
     sym,
@@ -256,7 +255,10 @@ impl<'a, 'genv, 'tcx: 'genv> RustItemCtxt<'a, 'genv, 'tcx> {
                     }
                     surface::GenericParamKind::Base => fhir::GenericParamKind::Base,
                 };
-                surface_params.insert(def_id, fhir::GenericParam { def_id, kind });
+                surface_params.insert(
+                    def_id,
+                    fhir::GenericParam { def_id, name: ParamName::Plain(param.name), kind },
+                );
             }
         }
 

--- a/crates/flux-desugar/src/desugar.rs
+++ b/crates/flux-desugar/src/desugar.rs
@@ -537,6 +537,7 @@ impl<'a, 'genv, 'tcx: 'genv> RustItemCtxt<'a, 'genv, 'tcx> {
         &mut self,
         fn_spec: &surface::FnSpec,
     ) -> Result<(fhir::Generics<'genv>, fhir::FnSig<'genv>)> {
+        let mut header = self.as_lift_cx().lift_fn_header();
         let (generics, decl) = if let Some(fn_sig) = &fn_spec.fn_sig {
             self.fn_sig_scope = Some(fn_sig.node_id);
 
@@ -565,6 +566,10 @@ impl<'a, 'genv, 'tcx: 'genv> RustItemCtxt<'a, 'genv, 'tcx> {
                 span: fn_sig.span,
                 lifted: false,
             };
+            // Fix up the span in asyncness
+            if let surface::Async::Yes { span, .. } = fn_sig.asyncness {
+                header.asyncness = hir::IsAsync::Async(span)
+            }
             (generics, decl)
         } else {
             (self.as_lift_cx().lift_generics()?, self.as_lift_cx().lift_fn_decl()?)
@@ -577,6 +582,7 @@ impl<'a, 'genv, 'tcx: 'genv> RustItemCtxt<'a, 'genv, 'tcx> {
         Ok((
             generics,
             fhir::FnSig {
+                header,
                 qualifiers: self.genv.alloc_slice(qual_names),
                 decl: self.genv.alloc(decl),
             },

--- a/crates/flux-fhir-analysis/locales/en-US.ftl
+++ b/crates/flux-fhir-analysis/locales/en-US.ftl
@@ -103,7 +103,7 @@ fhir_analysis_refined_unrefinable_type =
 fhir_analysis_incompatible_refinement =
     {$def_descr} has an incompatible refinement annotation
     .label = expected a refinement of `{$expected_ty}`
-    .expected_label = corresponding to the unrefined definition
+    .expected_label = unrefined {$def_descr} defined here
 
 fhir_analysis_incompatible_param_count =
     {$def_descr} has an incompatible refinement annotation

--- a/crates/flux-fhir-analysis/src/conv/mod.rs
+++ b/crates/flux-fhir-analysis/src/conv/mod.rs
@@ -344,17 +344,7 @@ pub(crate) fn conv_fn_decl(
     let mut env = Env::new(genv, generics.refinement_params, wfckresults)?;
     env.push_layer(Layer::list(&cx, late_bound_regions.len() as u32, &[])?);
 
-    let mut requires = vec![];
-    for req in decl.requires {
-        requires.push(cx.conv_requires(&mut env, req)?);
-    }
-
-    let mut inputs = vec![];
-    for ty in decl.inputs {
-        inputs.push(cx.conv_ty(&mut env, ty)?);
-    }
-
-    let output = cx.conv_fn_output(&mut env, &decl.output)?;
+    let fn_sig = cx.conv_fn_decl(&mut env, decl)?;
 
     let vars = late_bound_regions
         .iter()
@@ -362,10 +352,10 @@ pub(crate) fn conv_fn_decl(
         .cloned()
         .collect();
 
-    let fn_sig = rty::PolyFnSig::new(rty::FnSig::new(requires.into(), inputs.into(), output), vars);
-    let fn_sig = struct_compat::fn_sig(genv, decl, &fn_sig, def_id)?;
+    let poly_fn_sig = rty::PolyFnSig::new(fn_sig, vars);
+    let poly_fn_sig = struct_compat::fn_sig(genv, decl, &poly_fn_sig, def_id)?;
 
-    Ok(rty::EarlyBinder(fn_sig))
+    Ok(rty::EarlyBinder(poly_fn_sig))
 }
 
 pub(crate) fn conv_assoc_reft_def(
@@ -715,6 +705,22 @@ impl<'a, 'genv, 'tcx> ConvCtxt<'a, 'genv, 'tcx> {
         }
     }
 
+    fn conv_fn_decl(&mut self, env: &mut Env, decl: &fhir::FnDecl) -> QueryResult<rty::FnSig> {
+        let mut requires = vec![];
+        for req in decl.requires {
+            requires.push(self.conv_requires(env, req)?);
+        }
+
+        let mut inputs = vec![];
+        for ty in decl.inputs {
+            inputs.push(self.conv_ty(env, ty)?);
+        }
+
+        let output = self.conv_fn_output(env, &decl.output)?;
+
+        Ok(rty::FnSig::new(requires.into(), inputs.into(), output))
+    }
+
     fn conv_requires(
         &mut self,
         env: &mut Env,
@@ -802,6 +808,14 @@ impl<'a, 'genv, 'tcx> ConvCtxt<'a, 'genv, 'tcx> {
             fhir::TyKind::Ref(lft, fhir::MutTy { ty, mutbl }) => {
                 let region = self.conv_lifetime(env, *lft);
                 Ok(rty::Ty::mk_ref(region, self.conv_ty(env, ty)?, *mutbl))
+            }
+            fhir::TyKind::BareFn(bare_fn) => {
+                let mut env = Env::empty();
+                env.push_layer(Layer::list(&self, bare_fn.generic_params.len() as u32, &[])?);
+                let fn_sig = self.conv_fn_decl(&mut env, bare_fn.decl)?;
+                let poly_fn_sig = rty::Binder::new(fn_sig)
+                BaseTy::FnPtr()
+                todo!()
             }
             fhir::TyKind::Tuple(tys) => {
                 let tys: List<rty::Ty> =
@@ -1425,7 +1439,7 @@ impl<'a, 'genv, 'tcx> ConvCtxt<'a, 'genv, 'tcx> {
 }
 
 impl Env {
-    pub(crate) fn new(
+    fn new(
         genv: GlobalEnv,
         early_bound: &[fhir::RefineParam],
         wfckresults: &WfckResults,
@@ -1438,6 +1452,10 @@ impl Env {
             })
             .try_collect()?;
         Ok(Self { layers: vec![], early_bound })
+    }
+
+    fn empty() -> Self {
+        Self { layers: vec![], early_bound: Default::default() }
     }
 
     fn depth(&self) -> usize {

--- a/crates/flux-fhir-analysis/src/conv/mod.rs
+++ b/crates/flux-fhir-analysis/src/conv/mod.rs
@@ -34,7 +34,7 @@ use rustc_hir::{
     self as hir,
     def::DefKind,
     def_id::{DefId, LocalDefId},
-    PrimTy,
+    PrimTy, Safety,
 };
 use rustc_middle::{
     middle::resolve_bound_vars::ResolvedArg,
@@ -44,6 +44,7 @@ use rustc_span::{
     symbol::{kw, Ident},
     ErrorGuaranteed, Span, Symbol, DUMMY_SP,
 };
+use rustc_target::spec::abi;
 use rustc_trait_selection::traits;
 use rustc_type_ir::DebruijnIndex;
 
@@ -329,12 +330,15 @@ pub(crate) fn conv_qualifier(
     Ok(rty::Qualifier { name: qualifier.name, body, global: qualifier.global })
 }
 
-pub(crate) fn conv_fn_decl(
+pub(crate) fn conv_fn_sig(
     genv: GlobalEnv,
     def_id: MaybeExternId,
-    decl: &fhir::FnDecl,
+    fn_sig: &fhir::FnSig,
     wfckresults: &WfckResults,
 ) -> QueryResult<rty::EarlyBinder<rty::PolyFnSig>> {
+    let decl = &fn_sig.decl;
+    let header = fn_sig.header;
+
     let mut cx = ConvCtxt::new(genv, wfckresults);
 
     let late_bound_regions =
@@ -344,7 +348,7 @@ pub(crate) fn conv_fn_decl(
     let mut env = Env::new(genv, generics.refinement_params, wfckresults)?;
     env.push_layer(Layer::list(&cx, late_bound_regions.len() as u32, &[])?);
 
-    let fn_sig = cx.conv_fn_decl(&mut env, decl)?;
+    let fn_sig = cx.conv_fn_decl(&mut env, header.safety, header.abi, decl)?;
 
     let vars = late_bound_regions
         .iter()
@@ -705,7 +709,13 @@ impl<'a, 'genv, 'tcx> ConvCtxt<'a, 'genv, 'tcx> {
         }
     }
 
-    fn conv_fn_decl(&mut self, env: &mut Env, decl: &fhir::FnDecl) -> QueryResult<rty::FnSig> {
+    fn conv_fn_decl(
+        &mut self,
+        env: &mut Env,
+        safety: Safety,
+        abi: abi::Abi,
+        decl: &fhir::FnDecl,
+    ) -> QueryResult<rty::FnSig> {
         let mut requires = vec![];
         for req in decl.requires {
             requires.push(self.conv_requires(env, req)?);
@@ -718,7 +728,7 @@ impl<'a, 'genv, 'tcx> ConvCtxt<'a, 'genv, 'tcx> {
 
         let output = self.conv_fn_output(env, &decl.output)?;
 
-        Ok(rty::FnSig::new(requires.into(), inputs.into(), output))
+        Ok(rty::FnSig::new(safety, abi, requires.into(), inputs.into(), output))
     }
 
     fn conv_requires(
@@ -812,7 +822,8 @@ impl<'a, 'genv, 'tcx> ConvCtxt<'a, 'genv, 'tcx> {
             fhir::TyKind::BareFn(bare_fn) => {
                 let mut env = Env::empty();
                 env.push_layer(Layer::list(&self, bare_fn.generic_params.len() as u32, &[])?);
-                let fn_sig = self.conv_fn_decl(&mut env, bare_fn.decl)?;
+                let fn_sig =
+                    self.conv_fn_decl(&mut env, bare_fn.safety, bare_fn.abi, bare_fn.decl)?;
                 let vars = bare_fn
                     .generic_params
                     .iter()

--- a/crates/flux-fhir-analysis/src/conv/mod.rs
+++ b/crates/flux-fhir-analysis/src/conv/mod.rs
@@ -38,7 +38,7 @@ use rustc_hir::{
 };
 use rustc_middle::{
     middle::resolve_bound_vars::ResolvedArg,
-    ty::{self, AssocItem, AssocKind, BoundVar},
+    ty::{self, AssocItem, AssocKind, BoundRegionKind::BrNamed, BoundVar},
 };
 use rustc_span::{
     symbol::{kw, Ident},
@@ -498,7 +498,7 @@ impl<'a, 'genv, 'tcx> ConvCtxt<'a, 'genv, 'tcx> {
         let tcx = self.genv.tcx();
 
         let candidate = self.probe_single_bound_for_assoc_item(
-            || traits::supertraits(tcx, poly_trait_ref.to_rustc(tcx)),
+            || traits::supertraits(tcx, poly_trait_ref.to_rustc(tcx, rty::TraitRef::to_rustc)),
             constraint.ident,
         )?;
         let assoc_item_id = self
@@ -813,9 +813,13 @@ impl<'a, 'genv, 'tcx> ConvCtxt<'a, 'genv, 'tcx> {
                 let mut env = Env::empty();
                 env.push_layer(Layer::list(&self, bare_fn.generic_params.len() as u32, &[])?);
                 let fn_sig = self.conv_fn_decl(&mut env, bare_fn.decl)?;
-                let poly_fn_sig = rty::Binder::new(fn_sig)
-                BaseTy::FnPtr()
-                todo!()
+                let vars = bare_fn
+                    .generic_params
+                    .iter()
+                    .map(|param| self.param_as_bound_var(param))
+                    .try_collect()?;
+                let poly_fn_sig = rty::Binder::new(fn_sig, vars);
+                Ok(rty::BaseTy::FnPtr(poly_fn_sig).to_ty())
             }
             fhir::TyKind::Tuple(tys) => {
                 let tys: List<rty::Ty> =
@@ -1281,7 +1285,25 @@ impl<'a, 'genv, 'tcx> ConvCtxt<'a, 'genv, 'tcx> {
         Ok(rty::Binder::with_sort(rty::Ty::indexed(bty, rty::Expr::nu()), sort))
     }
 
-    pub fn conv_generic_args(
+    fn param_as_bound_var(
+        &mut self,
+        param: &fhir::GenericParam,
+    ) -> QueryResult<rty::BoundVariableKind> {
+        let def_id = param.def_id.to_def_id();
+        let name = self.genv.tcx().item_name(def_id);
+        match param.kind {
+            fhir::GenericParamKind::Lifetime => {
+                Ok(rty::BoundVariableKind::Region(BrNamed(def_id, name)))
+            }
+            fhir::GenericParamKind::Const { .. }
+            | fhir::GenericParamKind::Type { .. }
+            | fhir::GenericParamKind::Base => {
+                Err(query_bug!(param.def_id, "unsupported param kind `{:?}`", param.kind))
+            }
+        }
+    }
+
+    fn conv_generic_args(
         &mut self,
         env: &mut Env,
         def_id: DefId,

--- a/crates/flux-fhir-analysis/src/conv/mod.rs
+++ b/crates/flux-fhir-analysis/src/conv/mod.rs
@@ -821,7 +821,7 @@ impl<'a, 'genv, 'tcx> ConvCtxt<'a, 'genv, 'tcx> {
             }
             fhir::TyKind::BareFn(bare_fn) => {
                 let mut env = Env::empty();
-                env.push_layer(Layer::list(&self, bare_fn.generic_params.len() as u32, &[])?);
+                env.push_layer(Layer::list(self, bare_fn.generic_params.len() as u32, &[])?);
                 let fn_sig =
                     self.conv_fn_decl(&mut env, bare_fn.safety, bare_fn.abi, bare_fn.decl)?;
                 let vars = bare_fn

--- a/crates/flux-fhir-analysis/src/conv/struct_compat.rs
+++ b/crates/flux-fhir-analysis/src/conv/struct_compat.rs
@@ -56,9 +56,9 @@ pub(crate) fn fn_sig(
     let expected = Refiner::default(genv, &generics).refine_poly_fn_sig(&rust_fn_sig)?;
 
     let mut zipper = Zipper::new(genv, def_id);
-    zipper.enter_binders(fn_sig, &expected, |zipper, fn_sig, expected| {
-        zipper.zip_fn_sig(decl, fn_sig, expected);
-    });
+    if let Err(err) = zipper.zip_poly_fn_sig(fn_sig, &expected) {
+        zipper.emit_fn_sig_err(err, decl);
+    }
 
     zipper.errors.into_result()?;
 
@@ -155,6 +155,10 @@ impl<'genv, 'tcx> Zipper<'genv, 'tcx> {
         }
     }
 
+    fn zip_poly_fn_sig(&mut self, a: &rty::PolyFnSig, b: &rty::PolyFnSig) -> Result<(), FnSigErr> {
+        self.enter_binders(a, b, |this, a, b| this.zip_fn_sig(a, b))
+    }
+
     fn zip_variant(&mut self, a: &rty::PolyVariant, b: &rty::PolyVariant, variant_idx: VariantIdx) {
         self.enter_binders(a, b, |this, a, b| {
             // The args are always `GenericArgs::identity_for_item` inside the `EarlyBinder`
@@ -183,57 +187,34 @@ impl<'genv, 'tcx> Zipper<'genv, 'tcx> {
         });
     }
 
-    fn zip_fn_sig(&mut self, decl: &fhir::FnDecl, a: &rty::FnSig, b: &rty::FnSig) {
+    fn zip_fn_sig(&mut self, a: &rty::FnSig, b: &rty::FnSig) -> Result<(), FnSigErr> {
         if a.inputs().len() != b.inputs().len() {
-            self.errors
-                .emit(errors::IncompatibleParamCount::new(self.genv, decl, self.owner_id));
-            return;
+            Err(FnSigErr::ArgCountMismatch)?;
         }
         for (i, (ty_a, ty_b)) in iter::zip(a.inputs(), b.inputs()).enumerate() {
-            if self.zip_ty(ty_a, ty_b).is_err() {
-                self.errors.emit(errors::IncompatibleRefinement::fn_input(
-                    self.genv,
-                    self.owner_id,
-                    decl,
-                    i,
-                ));
-            }
+            self.zip_ty(ty_a, ty_b).map_err(|_| FnSigErr::FnInput(i))?;
         }
         self.enter_binders(a.output(), b.output(), |this, output_a, output_b| {
-            this.zip_output(decl, output_a, output_b);
-        });
+            this.zip_output(output_a, output_b)
+        })
     }
 
-    fn zip_output(&mut self, decl: &fhir::FnDecl, a: &rty::FnOutput, b: &rty::FnOutput) {
-        if self.zip_ty(&a.ret, &b.ret).is_err() {
-            self.errors.emit(errors::IncompatibleRefinement::fn_output(
-                self.genv,
-                self.owner_id,
-                decl,
-            ));
-        }
-        // Skip ensure clauses if there are any errors because we may not have a type for all locations
-        if self.errors.has_errors() {
-            return;
-        }
+    fn zip_output(&mut self, a: &rty::FnOutput, b: &rty::FnOutput) -> Result<(), FnSigErr> {
+        self.zip_ty(&a.ret, &b.ret)
+            .map_err(|_| FnSigErr::FnOutput)?;
+
         for (i, ensures) in a.ensures.iter().enumerate() {
             if let rty::Ensures::Type(path, ty_a) = ensures {
                 let loc = path.to_loc().unwrap();
                 let ty_b = self.locs.get(&loc).unwrap().clone();
-                if self.zip_ty(ty_a, &ty_b).is_err() {
-                    self.errors.emit(errors::IncompatibleRefinement::ensures(
-                        self.genv,
-                        self.owner_id,
-                        decl,
-                        &ty_b,
-                        i,
-                    ));
-                }
+                self.zip_ty(ty_a, &ty_b)
+                    .map_err(|_| FnSigErr::Ensures { i, expected: ty_b })?;
             }
         }
+        Ok(())
     }
 
-    fn zip_ty(&mut self, a: &rty::Ty, b: &rty::Ty) -> Result<(), Error> {
+    fn zip_ty(&mut self, a: &rty::Ty, b: &rty::Ty) -> Result<(), Mismatch> {
         match (a.kind(), b.kind()) {
             (rty::TyKind::Infer(vid), _) if a != &rty::Ty::trait_object_dummy_self() => {
                 let b = self.adjust_binders(b);
@@ -280,11 +261,11 @@ impl<'genv, 'tcx> Zipper<'genv, 'tcx> {
             ) => {
                 bug!("unexpected type {a:?}");
             }
-            _ => Err(Error::Incompatible),
+            _ => Err(Mismatch),
         }
     }
 
-    fn zip_bty(&mut self, a: &rty::BaseTy, b: &rty::BaseTy) -> Result<(), Error> {
+    fn zip_bty(&mut self, a: &rty::BaseTy, b: &rty::BaseTy) -> Result<(), Mismatch> {
         match (a, b) {
             (rty::BaseTy::Int(ity_a), rty::BaseTy::Int(ity_b)) => {
                 assert_eq_or_incompatible(ity_a, ity_b)
@@ -316,6 +297,10 @@ impl<'genv, 'tcx> Zipper<'genv, 'tcx> {
                 self.zip_region(re_a, re_b);
                 self.zip_ty(ty_a, ty_b)
             }
+            (rty::BaseTy::FnPtr(poly_sig_a), rty::BaseTy::FnPtr(poly_sig_b)) => {
+                self.zip_poly_fn_sig(poly_sig_a, poly_sig_b)
+                    .map_err(|_| Mismatch)
+            }
             (rty::BaseTy::Tuple(tys_a), rty::BaseTy::Tuple(tys_b)) => {
                 assert_eq_or_incompatible(tys_a.len(), tys_b.len())?;
                 for (ty_a, ty_b) in iter::zip(tys_a, tys_b) {
@@ -342,11 +327,15 @@ impl<'genv, 'tcx> Zipper<'genv, 'tcx> {
             (rty::BaseTy::Closure(..) | rty::BaseTy::Coroutine(..), _) => {
                 bug!("unexpected type `{a:?}`");
             }
-            _ => Err(Error::Incompatible),
+            _ => Err(Mismatch),
         }
     }
 
-    fn zip_generic_arg(&mut self, a: &rty::GenericArg, b: &rty::GenericArg) -> Result<(), Error> {
+    fn zip_generic_arg(
+        &mut self,
+        a: &rty::GenericArg,
+        b: &rty::GenericArg,
+    ) -> Result<(), Mismatch> {
         match (a, b) {
             (rty::GenericArg::Ty(ty_a), rty::GenericArg::Ty(ty_b)) => self.zip_ty(ty_a, ty_b),
             (rty::GenericArg::Base(ctor_a), rty::GenericArg::Base(ctor_b)) => {
@@ -361,11 +350,11 @@ impl<'genv, 'tcx> Zipper<'genv, 'tcx> {
             (rty::GenericArg::Const(ct_a), rty::GenericArg::Const(ct_b)) => {
                 self.zip_const(ct_a, ct_b)
             }
-            _ => Err(Error::Incompatible),
+            _ => Err(Mismatch),
         }
     }
 
-    fn zip_const(&mut self, a: &rty::Const, b: &ty::Const) -> Result<(), Error> {
+    fn zip_const(&mut self, a: &rty::Const, b: &ty::Const) -> Result<(), Mismatch> {
         match (&a.kind, &b.kind) {
             (rty::ConstKind::Infer(ty::InferConst::Var(cid)), _) => {
                 self.holes.consts.insert(*cid, b.clone());
@@ -381,7 +370,7 @@ impl<'genv, 'tcx> Zipper<'genv, 'tcx> {
             (rty::ConstKind::Unevaluated(c1), ty::ConstKind::Unevaluated(c2)) => {
                 assert_eq_or_incompatible(c1, c2)
             }
-            _ => Err(Error::Incompatible),
+            _ => Err(Mismatch),
         }
     }
 
@@ -396,7 +385,7 @@ impl<'genv, 'tcx> Zipper<'genv, 'tcx> {
         &mut self,
         a: &rty::Binder<rty::ExistentialPredicate>,
         b: &rty::Binder<rty::ExistentialPredicate>,
-    ) -> Result<(), Error> {
+    ) -> Result<(), Mismatch> {
         self.enter_binders(a, b, |this, a, b| {
             match (a, b) {
                 (
@@ -425,7 +414,7 @@ impl<'genv, 'tcx> Zipper<'genv, 'tcx> {
                     rty::ExistentialPredicate::AutoTrait(def_id_a),
                     rty::ExistentialPredicate::AutoTrait(def_id_b),
                 ) => assert_eq_or_incompatible(def_id_a, def_id_b),
-                _ => Err(Error::Incompatible),
+                _ => Err(Mismatch),
             }
         })
     }
@@ -468,17 +457,58 @@ impl<'genv, 'tcx> Zipper<'genv, 'tcx> {
             t.shift_out_escaping(self.b_index.as_u32() - self.a_index.as_u32())
         }
     }
+
+    fn emit_fn_sig_err(&mut self, err: FnSigErr, decl: &fhir::FnDecl) {
+        match err {
+            FnSigErr::ArgCountMismatch => {
+                self.errors.emit(errors::IncompatibleParamCount::new(
+                    self.genv,
+                    decl,
+                    self.owner_id,
+                ));
+            }
+            FnSigErr::FnInput(i) => {
+                self.errors.emit(errors::IncompatibleRefinement::fn_input(
+                    self.genv,
+                    self.owner_id,
+                    decl,
+                    i,
+                ));
+            }
+            FnSigErr::FnOutput => {
+                self.errors.emit(errors::IncompatibleRefinement::fn_output(
+                    self.genv,
+                    self.owner_id,
+                    decl,
+                ));
+            }
+            FnSigErr::Ensures { i, expected } => {
+                self.errors.emit(errors::IncompatibleRefinement::ensures(
+                    self.genv,
+                    self.owner_id,
+                    decl,
+                    &expected,
+                    i,
+                ));
+            }
+        }
+    }
 }
 
-fn assert_eq_or_incompatible<T: Eq>(a: T, b: T) -> Result<(), Error> {
+fn assert_eq_or_incompatible<T: Eq>(a: T, b: T) -> Result<(), Mismatch> {
     if a != b {
-        return Err(Error::Incompatible);
+        return Err(Mismatch);
     }
     Ok(())
 }
 
-enum Error {
-    Incompatible,
+struct Mismatch;
+
+enum FnSigErr {
+    ArgCountMismatch,
+    FnInput(usize),
+    FnOutput,
+    Ensures { i: usize, expected: rty::Ty },
 }
 
 mod errors {

--- a/crates/flux-fhir-analysis/src/lib.rs
+++ b/crates/flux-fhir-analysis/src/lib.rs
@@ -8,6 +8,7 @@ extern crate rustc_hir;
 extern crate rustc_hir_pretty;
 extern crate rustc_middle;
 extern crate rustc_span;
+extern crate rustc_target;
 extern crate rustc_trait_selection;
 extern crate rustc_type_ir;
 
@@ -407,7 +408,7 @@ fn fn_sig(genv: GlobalEnv, def_id: LocalDefId) -> QueryResult<rty::EarlyBinder<r
     let fn_sig = genv.desugar(def_id.local_id())?.fn_sig().unwrap();
     let wfckresults = genv.check_wf(def_id.local_id())?;
     let defns = genv.spec_func_defns()?;
-    let fn_sig = conv::conv_fn_decl(genv, def_id, fn_sig.decl, &wfckresults)?
+    let fn_sig = conv::conv_fn_sig(genv, def_id, fn_sig, &wfckresults)?
         .map(|fn_sig| fn_sig.normalize(defns));
 
     if config::dump_rty() {

--- a/crates/flux-middle/src/fhir.rs
+++ b/crates/flux-middle/src/fhir.rs
@@ -353,7 +353,7 @@ pub struct TyAlias<'fhir> {
     pub span: Span,
     /// Whether this alias was [lifted] from a `hir` alias
     ///
-    /// [lifted]: lift::lift_type_alias
+    /// [lifted]: lift::LiftCtxt::lift_type_alias
     pub lifted: bool,
 }
 

--- a/crates/flux-middle/src/fhir.rs
+++ b/crates/flux-middle/src/fhir.rs
@@ -29,7 +29,7 @@ pub use rustc_hir::PrimTy;
 use rustc_hir::{
     def::DefKind,
     def_id::{DefId, LocalDefId},
-    ItemId, OwnerId,
+    ItemId, OwnerId, ParamName,
 };
 use rustc_index::newtype_index;
 use rustc_macros::{Decodable, Encodable, TyDecodable, TyEncodable};
@@ -83,6 +83,7 @@ pub struct Generics<'fhir> {
 #[derive(Debug, Clone, Copy)]
 pub struct GenericParam<'fhir> {
     pub def_id: LocalDefId,
+    pub name: ParamName,
     pub kind: GenericParamKind<'fhir>,
 }
 
@@ -407,9 +408,7 @@ pub struct VariantRet<'fhir> {
 
 #[derive(Clone, Copy)]
 pub struct FnDecl<'fhir> {
-    /// example: vec![(0 <= n), (l: i32)]
     pub requires: &'fhir [Requires<'fhir>],
-    /// example: vec![(x: StrRef(l))]
     pub inputs: &'fhir [Ty<'fhir>],
     pub output: FnOutput<'fhir>,
     pub span: Span,
@@ -473,6 +472,7 @@ pub enum TyKind<'fhir> {
     Constr(Expr<'fhir>, &'fhir Ty<'fhir>),
     StrgRef(Lifetime, &'fhir PathExpr<'fhir>, &'fhir Ty<'fhir>),
     Ref(Lifetime, MutTy<'fhir>),
+    BareFn(&'fhir BareFnTy<'fhir>),
     Tuple(&'fhir [Ty<'fhir>]),
     Array(&'fhir Ty<'fhir>, ConstArg),
     RawPtr(&'fhir Ty<'fhir>, Mutability),
@@ -480,6 +480,12 @@ pub enum TyKind<'fhir> {
     TraitObject(&'fhir [PolyTraitRef<'fhir>], Lifetime, TraitObjectSyntax),
     Never,
     Infer,
+}
+
+pub struct BareFnTy<'fhir> {
+    pub generic_params: &'fhir [GenericParam<'fhir>],
+    pub decl: &'fhir FnDecl<'fhir>,
+    pub param_names: &'fhir [Ident],
 }
 
 #[derive(Clone, Copy)]
@@ -1068,7 +1074,7 @@ impl<'fhir> Generics<'fhir> {
             } else {
                 param.kind
             };
-            GenericParam { def_id: param.def_id, kind }
+            GenericParam { kind, ..*param }
         }));
         Generics { params, ..self }
     }
@@ -1217,6 +1223,9 @@ impl fmt::Debug for Ty<'_> {
             TyKind::Ref(_lft, mut_ty) => {
                 write!(f, "&{}{:?}", mut_ty.mutbl.prefix_str(), mut_ty.ty)
             }
+            TyKind::BareFn(bare_fn_ty) => {
+                write!(f, "{bare_fn_ty:?}")
+            }
             TyKind::Tuple(tys) => write!(f, "({:?})", tys.iter().format(", ")),
             TyKind::Array(ty, len) => write!(f, "[{ty:?}; {len:?}]"),
             TyKind::Never => write!(f, "!"),
@@ -1234,6 +1243,22 @@ impl fmt::Debug for Ty<'_> {
                 write!(f, "dyn {poly_traits:?}")
             }
         }
+    }
+}
+
+impl fmt::Debug for BareFnTy<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if !self.generic_params.is_empty() {
+            write!(
+                f,
+                "for<{}>",
+                self.generic_params
+                    .iter()
+                    .map(|param| param.name.ident())
+                    .format(",")
+            )?;
+        }
+        write!(f, "{:?}", self.decl)
     }
 }
 

--- a/crates/flux-middle/src/fhir.rs
+++ b/crates/flux-middle/src/fhir.rs
@@ -495,7 +495,7 @@ pub struct MutTy<'fhir> {
 }
 
 /// Our surface syntax doesn't have lifetimes. To deal with them we create a *hole* for every lifetime
-/// which we then resolve during `annot_check` when zipping against the lifted version.
+/// which we then resolve when we check for structural compatibility against the rust type.
 #[derive(Copy, Clone, PartialEq, Eq)]
 pub enum Lifetime {
     /// A lifetime hole created during desugaring.
@@ -733,7 +733,7 @@ pub enum ParamKind {
     /// A location declared with `x: &strg T` syntax.
     Loc,
     /// A parameter introduced with `x: T` syntax that we know *syntactically* is always and error
-    /// to used inside a refinement. For example, consider the following:
+    /// to use inside a refinement. For example, consider the following:
     /// ```ignore
     /// fn(x: {v. i32[v] | v > 0}) -> i32[x]
     /// ```
@@ -1136,18 +1136,6 @@ impl fmt::Debug for FnSig<'_> {
 
 impl fmt::Debug for FnDecl<'_> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        // if !self.generics.refinement_params.is_empty() {
-        //     write!(
-        //         f,
-        //         "for<{}> ",
-        //         self.generics
-        //             .refinement_params
-        //             .iter()
-        //             .format_with(", ", |param, f| {
-        //                 f(&format_args!("{}: {:?}", param.name, param.sort))
-        //             })
-        //     )?;
-        // }
         if !self.requires.is_empty() {
             write!(f, "[{:?}] ", self.requires.iter().format(", "))?;
         }

--- a/crates/flux-middle/src/fhir.rs
+++ b/crates/flux-middle/src/fhir.rs
@@ -29,7 +29,7 @@ pub use rustc_hir::PrimTy;
 use rustc_hir::{
     def::DefKind,
     def_id::{DefId, LocalDefId},
-    ItemId, OwnerId, ParamName,
+    FnHeader, ItemId, OwnerId, ParamName, Safety,
 };
 use rustc_index::newtype_index;
 use rustc_macros::{Decodable, Encodable, TyDecodable, TyEncodable};
@@ -37,6 +37,7 @@ pub use rustc_middle::mir::Mutability;
 use rustc_middle::{middle::resolve_bound_vars::ResolvedArg, ty::TyCtxt};
 use rustc_span::{symbol::Ident, Span, Symbol};
 pub use rustc_target::abi::VariantIdx;
+use rustc_target::spec::abi;
 
 use crate::{global_env::GlobalEnv, pretty, MaybeExternId};
 
@@ -428,6 +429,7 @@ pub struct Requires<'fhir> {
 
 #[derive(Clone, Copy)]
 pub struct FnSig<'fhir> {
+    pub header: FnHeader,
     //// List of local qualifiers for this function
     pub qualifiers: &'fhir [Ident],
     pub decl: &'fhir FnDecl<'fhir>,
@@ -483,6 +485,8 @@ pub enum TyKind<'fhir> {
 }
 
 pub struct BareFnTy<'fhir> {
+    pub safety: Safety,
+    pub abi: abi::Abi,
     pub generic_params: &'fhir [GenericParam<'fhir>],
     pub decl: &'fhir FnDecl<'fhir>,
     pub param_names: &'fhir [Ident],

--- a/crates/flux-middle/src/fhir/lift.rs
+++ b/crates/flux-middle/src/fhir/lift.rs
@@ -368,7 +368,7 @@ impl<'a, 'genv, 'tcx> LiftCtxt<'a, 'genv, 'tcx> {
         let generic_params = try_alloc_slice!(self.genv, bare_fn.generic_params, |param| {
             self.lift_generic_param(param)
         })?;
-        let decl = self.lift_fn_decl_inner(span, &bare_fn.decl)?;
+        let decl = self.lift_fn_decl_inner(span, bare_fn.decl)?;
         Ok(fhir::BareFnTy {
             safety: bare_fn.safety,
             abi: bare_fn.abi,

--- a/crates/flux-middle/src/fhir/lift.rs
+++ b/crates/flux-middle/src/fhir/lift.rs
@@ -76,7 +76,7 @@ impl<'a, 'genv, 'tcx> LiftCtxt<'a, 'genv, 'tcx> {
                 fhir::GenericParamKind::Const { ty, is_host_effect }
             }
         };
-        Ok(fhir::GenericParam { def_id: param.def_id, kind })
+        Ok(fhir::GenericParam { def_id: param.def_id, name: param.name, kind })
     }
 
     fn lift_generics_inner(&mut self, generics: &hir::Generics) -> Result<fhir::Generics<'genv>> {

--- a/crates/flux-middle/src/fhir/lift.rs
+++ b/crates/flux-middle/src/fhir/lift.rs
@@ -5,8 +5,8 @@ use flux_errors::ErrorGuaranteed;
 use hir::OwnerId;
 use rustc_data_structures::unord::UnordMap;
 use rustc_errors::Diagnostic;
-use rustc_hir as hir;
-use rustc_hir::def_id::LocalDefId;
+use rustc_hir::{self as hir, def_id::LocalDefId, FnHeader};
+use rustc_span::Span;
 
 use super::{FhirId, FluxOwnerId};
 use crate::{
@@ -166,6 +166,16 @@ impl<'a, 'genv, 'tcx> LiftCtxt<'a, 'genv, 'tcx> {
         })
     }
 
+    pub fn lift_fn_header(&mut self) -> FnHeader {
+        let def_id = self.owner.def_id;
+        let hir_id = self.genv.tcx().local_def_id_to_hir_id(def_id);
+        self.genv
+            .hir()
+            .fn_sig_by_hir_id(hir_id)
+            .expect("item does not have a `FnDecl`")
+            .header
+    }
+
     pub fn lift_fn_decl(&mut self) -> Result<fhir::FnDecl<'genv>> {
         let def_id = self.owner.def_id;
         let hir_id = self.genv.tcx().local_def_id_to_hir_id(def_id);
@@ -175,15 +185,20 @@ impl<'a, 'genv, 'tcx> LiftCtxt<'a, 'genv, 'tcx> {
             .fn_sig_by_hir_id(hir_id)
             .expect("item does not have a `FnDecl`");
 
-        let inputs = try_alloc_slice!(self.genv, &fn_sig.decl.inputs, |ty| self.lift_ty(ty))?;
+        self.lift_fn_decl_inner(fn_sig.span, fn_sig.decl)
+    }
 
-        let output = fhir::FnOutput {
-            params: &[],
-            ensures: &[],
-            ret: self.lift_fn_ret_ty(&fn_sig.decl.output)?,
-        };
+    fn lift_fn_decl_inner(
+        &mut self,
+        span: Span,
+        decl: &hir::FnDecl,
+    ) -> Result<fhir::FnDecl<'genv>> {
+        let inputs = try_alloc_slice!(self.genv, &decl.inputs, |ty| self.lift_ty(ty))?;
 
-        Ok(fhir::FnDecl { requires: &[], inputs, output, span: fn_sig.span, lifted: true })
+        let output =
+            fhir::FnOutput { params: &[], ensures: &[], ret: self.lift_fn_ret_ty(&decl.output)? };
+
+        Ok(fhir::FnDecl { requires: &[], inputs, output, span, lifted: true })
     }
 
     fn lift_fn_ret_ty(&mut self, ret_ty: &hir::FnRetTy) -> Result<fhir::Ty<'genv>> {
@@ -293,6 +308,10 @@ impl<'a, 'genv, 'tcx> LiftCtxt<'a, 'genv, 'tcx> {
             hir::TyKind::Ref(lft, mut_ty) => {
                 fhir::TyKind::Ref(self.lift_lifetime(lft)?, self.lift_mut_ty(mut_ty)?)
             }
+            hir::TyKind::BareFn(bare_fn) => {
+                let bare_fn = self.lift_bare_fn(ty.span, bare_fn)?;
+                fhir::TyKind::BareFn(self.genv.alloc(bare_fn))
+            }
             hir::TyKind::Never => fhir::TyKind::Never,
             hir::TyKind::Tup(tys) => {
                 let tys = try_alloc_slice!(self.genv, tys, |ty| self.lift_ty(ty))?;
@@ -339,6 +358,24 @@ impl<'a, 'genv, 'tcx> LiftCtxt<'a, 'genv, 'tcx> {
             }
         };
         Ok(fhir::Ty { kind, span: ty.span })
+    }
+
+    fn lift_bare_fn(
+        &mut self,
+        span: Span,
+        bare_fn: &hir::BareFnTy,
+    ) -> Result<fhir::BareFnTy<'genv>> {
+        let generic_params = try_alloc_slice!(self.genv, bare_fn.generic_params, |param| {
+            self.lift_generic_param(param)
+        })?;
+        let decl = self.lift_fn_decl_inner(span, &bare_fn.decl)?;
+        Ok(fhir::BareFnTy {
+            safety: bare_fn.safety,
+            abi: bare_fn.abi,
+            generic_params,
+            decl: self.genv.alloc(decl),
+            param_names: self.genv.alloc_slice(bare_fn.param_names),
+        })
     }
 
     fn lift_lifetime(&self, lft: &hir::Lifetime) -> Result<fhir::Lifetime> {

--- a/crates/flux-middle/src/fhir/visit.rs
+++ b/crates/flux-middle/src/fhir/visit.rs
@@ -363,6 +363,7 @@ pub fn walk_ty<'v, V: Visitor<'v>>(vis: &mut V, ty: &Ty<'v>) {
             vis.visit_lifetime(&lft);
             vis.visit_ty(mty.ty);
         }
+        TyKind::BareFn(bare_fn) => vis.visit_fn_decl(bare_fn.decl),
         TyKind::Tuple(tys) => {
             walk_list!(vis, visit_ty, tys);
         }

--- a/crates/flux-middle/src/queries.rs
+++ b/crates/flux-middle/src/queries.rs
@@ -77,7 +77,7 @@ pub enum QueryErr {
     },
     /// Used to report bugs, typically this means executing an arm in a match we though it was
     /// unreachable. Use this instead of panicking if it is easy to return a [`QueryErr`]. Use
-    /// [`QueryErr::bug`] or [`query_bug!`] to construct this variant to track source location.
+    /// [`QueryErr::bug`] or [`crate::query_bug!`] to construct this variant to track source location.
     Bug {
         def_id: Option<DefId>,
         location: String,

--- a/crates/flux-middle/src/rty/fold.rs
+++ b/crates/flux-middle/src/rty/fold.rs
@@ -793,7 +793,7 @@ impl TypeFoldable for FnSig {
         let requires = self.requires.try_fold_with(folder)?;
         let args = self.inputs.try_fold_with(folder)?;
         let output = self.output.try_fold_with(folder)?;
-        Ok(FnSig::new(requires, args, output))
+        Ok(FnSig::new(self.safety, self.abi, requires, args, output))
     }
 }
 

--- a/crates/flux-middle/src/rty/fold.rs
+++ b/crates/flux-middle/src/rty/fold.rs
@@ -992,6 +992,7 @@ impl TypeSuperVisitable for BaseTy {
             BaseTy::Slice(ty) => ty.visit_with(visitor),
             BaseTy::RawPtr(ty, _) => ty.visit_with(visitor),
             BaseTy::Ref(_, ty, _) => ty.visit_with(visitor),
+            BaseTy::FnPtr(poly_fn_sig) => poly_fn_sig.visit_with(visitor),
             BaseTy::Tuple(tys) => tys.visit_with(visitor),
             BaseTy::Array(ty, _) => ty.visit_with(visitor),
             BaseTy::Coroutine(_, resume_ty, upvars) => {
@@ -1027,6 +1028,7 @@ impl TypeSuperFoldable for BaseTy {
             BaseTy::Ref(re, ty, mutbl) => {
                 BaseTy::Ref(re.try_fold_with(folder)?, ty.try_fold_with(folder)?, *mutbl)
             }
+            BaseTy::FnPtr(decl) => BaseTy::FnPtr(decl.try_fold_with(folder)?),
             BaseTy::Tuple(tys) => BaseTy::Tuple(tys.try_fold_with(folder)?),
             BaseTy::Array(ty, c) => {
                 BaseTy::Array(ty.try_fold_with(folder)?, c.try_fold_with(folder)?)

--- a/crates/flux-middle/src/rty/mod.rs
+++ b/crates/flux-middle/src/rty/mod.rs
@@ -948,14 +948,14 @@ pub struct EarlyBinder<T>(pub T);
 
 pub type PolyFnSig = Binder<FnSig>;
 
-#[derive(Clone, TyEncodable, TyDecodable)]
+#[derive(Clone, Eq, PartialEq, Hash, TyEncodable, TyDecodable)]
 pub struct FnSig {
     requires: List<Expr>,
     inputs: List<Ty>,
     output: Binder<FnOutput>,
 }
 
-#[derive(Clone, Debug, TyEncodable, TyDecodable)]
+#[derive(Clone, Eq, PartialEq, Hash, Debug, TyEncodable, TyDecodable)]
 pub struct FnOutput {
     pub ret: Ty,
     pub ensures: List<Ensures>,
@@ -1367,6 +1367,7 @@ pub enum BaseTy {
     Float(FloatTy),
     RawPtr(Ty, Mutability),
     Ref(Region, Ty, Mutability),
+    FnPtr(PolyFnSig),
     Tuple(List<Ty>),
     Array(Ty, Const),
     Never,
@@ -1527,6 +1528,7 @@ impl BaseTy {
             | BaseTy::Char
             | BaseTy::RawPtr(..)
             | BaseTy::Ref(..)
+            | BaseTy::FnPtr(..)
             | BaseTy::Tuple(_)
             | BaseTy::Array(_, _)
             | BaseTy::Closure(_, _)
@@ -1556,6 +1558,9 @@ impl BaseTy {
             BaseTy::RawPtr(ty, mutbl) => ty::Ty::new_ptr(tcx, ty.to_rustc(tcx), *mutbl),
             BaseTy::Ref(re, ty, mutbl) => {
                 ty::Ty::new_ref(tcx, re.to_rustc(tcx), ty.to_rustc(tcx), *mutbl)
+            }
+            BaseTy::FnPtr(poly_fn_sig) => {
+                todo!()
             }
             BaseTy::Tuple(tys) => {
                 let ts = tys.iter().map(|ty| ty.to_rustc(tcx)).collect_vec();

--- a/crates/flux-middle/src/rty/mod.rs
+++ b/crates/flux-middle/src/rty/mod.rs
@@ -24,7 +24,7 @@ use flux_common::bug;
 use itertools::Itertools;
 pub use normalize::SpecFuncDefns;
 use rustc_data_structures::unord::UnordMap;
-use rustc_hir::{def_id::DefId, LangItem};
+use rustc_hir::{def_id::DefId, LangItem, Safety};
 use rustc_index::{newtype_index, IndexSlice};
 use rustc_macros::{Decodable, Encodable, TyDecodable, TyEncodable};
 use rustc_middle::ty::{ParamConst, TyCtxt};
@@ -34,6 +34,7 @@ pub use rustc_middle::{
 };
 use rustc_span::{sym, symbol::kw, Symbol};
 pub use rustc_target::abi::{VariantIdx, FIRST_VARIANT};
+use rustc_target::spec::abi;
 pub use rustc_type_ir::{TyVid, INNERMOST};
 pub use SortInfer::*;
 
@@ -259,10 +260,6 @@ impl PolyTraitRef {
     pub fn def_id(&self) -> DefId {
         self.as_ref().skip_binder().def_id
     }
-
-    pub fn to_rustc<'tcx>(&self, tcx: TyCtxt<'tcx>) -> rustc_middle::ty::PolyTraitRef<'tcx> {
-        rustc_middle::ty::Binder::bind_with_vars(self.value.to_rustc(tcx), self.vars.to_rustc(tcx))
-    }
 }
 
 #[derive(Clone, PartialEq, Eq, Hash, TyEncodable, TyDecodable)]
@@ -296,13 +293,12 @@ impl ExistentialPredicate {
     }
 }
 
-impl PolyExistentialPredicate {
+impl ExistentialPredicate {
     pub fn to_rustc<'tcx>(
         &self,
         tcx: TyCtxt<'tcx>,
-    ) -> rustc_middle::ty::PolyExistentialPredicate<'tcx> {
-        assert!(self.vars.is_empty());
-        let pred = match &self.value {
+    ) -> rustc_middle::ty::ExistentialPredicate<'tcx> {
+        match self {
             ExistentialPredicate::Trait(trait_ref) => {
                 let trait_ref = rustc_middle::ty::ExistentialTraitRef {
                     def_id: trait_ref.def_id,
@@ -322,8 +318,7 @@ impl PolyExistentialPredicate {
             ExistentialPredicate::AutoTrait(def_id) => {
                 rustc_middle::ty::ExistentialPredicate::AutoTrait(*def_id)
             }
-        };
-        rustc_middle::ty::Binder::dummy(pred)
+        }
     }
 }
 
@@ -940,6 +935,19 @@ impl<T> Binder<T> {
             [BoundVariableKind::Refine(sort, ..)] => sort.clone(),
             _ => bug!("expected single-sorted binder"),
         }
+    }
+
+    pub fn to_rustc<'tcx, R>(
+        &self,
+        tcx: TyCtxt<'tcx>,
+        f: impl FnOnce(&T, TyCtxt<'tcx>) -> R,
+    ) -> rustc_middle::ty::Binder<'tcx, R>
+    where
+        R: rustc_middle::ty::TypeVisitable<TyCtxt<'tcx>>,
+    {
+        let vars = self.vars.to_rustc(tcx);
+        let value = f(&self.value, tcx);
+        rustc_middle::ty::Binder::bind_with_vars(value, vars)
     }
 }
 
@@ -1560,7 +1568,7 @@ impl BaseTy {
                 ty::Ty::new_ref(tcx, re.to_rustc(tcx), ty.to_rustc(tcx), *mutbl)
             }
             BaseTy::FnPtr(poly_fn_sig) => {
-                todo!()
+                ty::Ty::new_fn_ptr(tcx, poly_fn_sig.to_rustc(tcx, FnSig::to_rustc))
             }
             BaseTy::Tuple(tys) => {
                 let ts = tys.iter().map(|ty| ty.to_rustc(tcx)).collect_vec();
@@ -1578,7 +1586,7 @@ impl BaseTy {
             BaseTy::Dynamic(exi_preds, re) => {
                 let preds: Vec<_> = exi_preds
                     .iter()
-                    .map(|pred| pred.to_rustc(tcx))
+                    .map(|pred| pred.to_rustc(tcx, ExistentialPredicate::to_rustc))
                     .collect_vec();
                 let preds = tcx.mk_poly_existential_predicates(&preds);
                 ty::Ty::new_dynamic(tcx, preds, re.to_rustc(tcx), rustc_middle::ty::DynKind::Dyn)
@@ -2050,11 +2058,25 @@ impl FnSig {
     pub fn output(&self) -> &Binder<FnOutput> {
         &self.output
     }
+
+    fn to_rustc<'tcx>(&self, tcx: TyCtxt<'tcx>) -> rustc_middle::ty::FnSig<'tcx> {
+        tcx.mk_fn_sig(
+            self.inputs().iter().map(|ty| ty.to_rustc(tcx)),
+            self.output().as_ref().skip_binder().to_rustc(tcx),
+            false,
+            Safety::Safe,
+            abi::Abi::Rust,
+        )
+    }
 }
 
 impl FnOutput {
     pub fn new(ret: Ty, ensures: impl Into<List<Ensures>>) -> Self {
         Self { ret, ensures: ensures.into() }
+    }
+
+    fn to_rustc<'tcx>(&self, tcx: TyCtxt<'tcx>) -> rustc_middle::ty::Ty<'tcx> {
+        self.ret.to_rustc(tcx)
     }
 }
 

--- a/crates/flux-middle/src/rty/mod.rs
+++ b/crates/flux-middle/src/rty/mod.rs
@@ -386,6 +386,8 @@ impl FnTraitPredicate {
             .collect();
 
         let fn_sig = FnSig::new(
+            Safety::Safe,
+            abi::Abi::RustCall,
             List::empty(),
             inputs,
             Binder::new(FnOutput::new(self.output.clone(), vec![]), List::empty()),
@@ -958,6 +960,8 @@ pub type PolyFnSig = Binder<FnSig>;
 
 #[derive(Clone, Eq, PartialEq, Hash, TyEncodable, TyDecodable)]
 pub struct FnSig {
+    safety: Safety,
+    abi: abi::Abi,
     requires: List<Expr>,
     inputs: List<Ty>,
     output: Binder<FnOutput>,
@@ -1870,7 +1874,10 @@ impl CoroutineObligPredicate {
         let inputs = List::from_arr([env_ty, resume_ty.clone()]);
         let output = Binder::new(FnOutput::new(self.output.clone(), vec![]), List::empty());
 
-        PolyFnSig::new(FnSig::new(List::empty(), inputs, output), List::from(vars))
+        PolyFnSig::new(
+            FnSig::new(Safety::Safe, abi::Abi::RustCall, List::empty(), inputs, output),
+            List::from(vars),
+        )
     }
 }
 
@@ -2043,8 +2050,14 @@ impl VariantSig {
 }
 
 impl FnSig {
-    pub fn new(requires: List<Expr>, inputs: List<Ty>, output: Binder<FnOutput>) -> Self {
-        FnSig { requires, inputs, output }
+    pub fn new(
+        safety: Safety,
+        abi: abi::Abi,
+        requires: List<Expr>,
+        inputs: List<Ty>,
+        output: Binder<FnOutput>,
+    ) -> Self {
+        FnSig { safety, abi, requires, inputs, output }
     }
 
     pub fn requires(&self) -> &[Expr] {
@@ -2064,8 +2077,8 @@ impl FnSig {
             self.inputs().iter().map(|ty| ty.to_rustc(tcx)),
             self.output().as_ref().skip_binder().to_rustc(tcx),
             false,
-            Safety::Safe,
-            abi::Abi::Rust,
+            self.safety,
+            self.abi,
         )
     }
 }
@@ -2192,7 +2205,13 @@ impl EarlyBinder<PolyVariant> {
             poly_variant.as_ref().map(|variant| {
                 let ret = variant.ret().shift_in_escaping(1);
                 let output = Binder::new(FnOutput::new(ret, vec![]), List::empty());
-                FnSig::new(List::empty(), variant.fields.clone(), output)
+                FnSig::new(
+                    Safety::Safe,
+                    abi::Abi::Rust,
+                    List::empty(),
+                    variant.fields.clone(),
+                    output,
+                )
             })
         })
     }

--- a/crates/flux-middle/src/rty/pretty.rs
+++ b/crates/flux-middle/src/rty/pretty.rs
@@ -428,6 +428,9 @@ impl Pretty for BaseTy {
                 }
                 w!("{}{:?}",  ^mutbl.prefix_str(), ty)
             }
+            BaseTy::FnPtr(poly_fn_sig) => {
+                w!("{:?}", poly_fn_sig)
+            }
             BaseTy::Tuple(tys) => {
                 if let [ty] = &tys[..] {
                     w!("({:?},)", ty)

--- a/crates/flux-middle/src/rty/refining.rs
+++ b/crates/flux-middle/src/rty/refining.rs
@@ -280,7 +280,7 @@ impl<'genv, 'tcx> Refiner<'genv, 'tcx> {
                 .try_collect()?;
             let ret = self.refine_ty(fn_sig.output())?.shift_in_escaping(1);
             let output = rty::Binder::new(rty::FnOutput::new(ret, vec![]), List::empty());
-            Ok(rty::FnSig::new(List::empty(), inputs, output))
+            Ok(rty::FnSig::new(fn_sig.safety, fn_sig.abi, List::empty(), inputs, output))
         })
     }
 
@@ -415,7 +415,9 @@ impl<'genv, 'tcx> Refiner<'genv, 'tcx> {
             rustc::ty::TyKind::Str => rty::BaseTy::Str,
             rustc::ty::TyKind::Slice(ty) => rty::BaseTy::Slice(self.refine_ty(ty)?),
             rustc::ty::TyKind::Char => rty::BaseTy::Char,
-            rustc::ty::TyKind::FnPtr(_) => bug!("TODO: refine_ty: FnSig"),
+            rustc::ty::TyKind::FnPtr(poly_fn_sig) => {
+                rty::BaseTy::FnPtr(self.as_default().refine_poly_fn_sig(poly_fn_sig)?)
+            }
             rustc::ty::TyKind::RawPtr(ty, mu) => {
                 rty::BaseTy::RawPtr(self.as_default().refine_ty(ty)?, *mu)
             }

--- a/crates/flux-middle/src/rustc/lowering.rs
+++ b/crates/flux-middle/src/rustc/lowering.rs
@@ -649,7 +649,7 @@ pub fn lower_fn_sig<'tcx>(
                 .map(|ty| lower_ty(tcx, ty))
                 .try_collect()?,
         );
-        Ok(FnSig { inputs_and_output })
+        Ok(FnSig { safety: fn_sig.safety, abi: fn_sig.abi, inputs_and_output })
     })
 }
 

--- a/crates/flux-middle/src/rustc/ty.rs
+++ b/crates/flux-middle/src/rustc/ty.rs
@@ -6,7 +6,7 @@ use std::fmt;
 
 use flux_common::bug;
 use itertools::Itertools;
-use rustc_hir::def_id::DefId;
+use rustc_hir::{def_id::DefId, Safety};
 use rustc_index::{IndexSlice, IndexVec};
 use rustc_macros::{TyDecodable, TyEncodable};
 use rustc_middle::ty::{self as rustc_ty, AdtFlags, ParamConst, TyCtxt};
@@ -19,6 +19,7 @@ pub use rustc_middle::{
 };
 use rustc_span::{symbol::kw, Symbol};
 pub use rustc_target::abi::{FieldIdx, VariantIdx, FIRST_VARIANT};
+use rustc_target::spec::abi;
 pub use rustc_type_ir::InferConst;
 
 use self::subst::Subst;
@@ -115,6 +116,8 @@ pub struct ProjectionPredicate {
 }
 #[derive(Clone, Hash, PartialEq, Eq, TyEncodable, TyDecodable)]
 pub struct FnSig {
+    pub safety: Safety,
+    pub abi: abi::Abi,
     pub(crate) inputs_and_output: List<Ty>,
 }
 

--- a/crates/flux-middle/src/rustc/ty/subst.rs
+++ b/crates/flux-middle/src/rustc/ty/subst.rs
@@ -22,7 +22,11 @@ where
 
 impl Subst for FnSig {
     fn subst(&self, args: &[GenericArg]) -> Self {
-        FnSig { inputs_and_output: self.inputs_and_output.subst(args) }
+        FnSig {
+            safety: self.safety,
+            abi: self.abi,
+            inputs_and_output: self.inputs_and_output.subst(args),
+        }
     }
 }
 

--- a/crates/flux-middle/src/sort_of.rs
+++ b/crates/flux-middle/src/sort_of.rs
@@ -106,15 +106,16 @@ impl<'sess, 'tcx> GlobalEnv<'sess, 'tcx> {
         match &ty.kind {
             fhir::TyKind::BaseTy(bty) | fhir::TyKind::Indexed(bty, _) => self.sort_of_bty(bty),
             fhir::TyKind::Exists(_, ty) | fhir::TyKind::Constr(_, ty) => self.sort_of_ty(ty),
-            fhir::TyKind::RawPtr(_, _)
-            | fhir::TyKind::Ref(_, _)
+            fhir::TyKind::RawPtr(..)
+            | fhir::TyKind::Ref(..)
             | fhir::TyKind::Tuple(_)
-            | fhir::TyKind::Array(_, _)
-            | fhir::TyKind::TraitObject(_, _, _)
+            | fhir::TyKind::Array(..)
+            | fhir::TyKind::TraitObject(..)
+            | fhir::TyKind::BareFn(_)
             | fhir::TyKind::Never => Ok(Some(rty::Sort::unit())),
-            fhir::TyKind::Infer
-            | fhir::TyKind::StrgRef(..)
-            | fhir::TyKind::OpaqueDef(_, _, _, _) => Ok(None),
+            fhir::TyKind::Infer | fhir::TyKind::StrgRef(..) | fhir::TyKind::OpaqueDef(..) => {
+                Ok(None)
+            }
         }
     }
 

--- a/crates/flux-refineck/src/lib.rs
+++ b/crates/flux-refineck/src/lib.rs
@@ -111,6 +111,7 @@ pub fn check_fn(
         let errors = fcx.check(cache, cstr, config.scrape_quals).emit(&genv)?;
 
         tracing::info!("check_fn::fixpoint");
+        #[expect(clippy::collapsible_else_if, reason = "it looks better")]
         if genv.should_fail(def_id) {
             if errors.is_empty() {
                 report_expected_neg(genv, def_id)

--- a/crates/flux-refineck/src/type_env.rs
+++ b/crates/flux-refineck/src/type_env.rs
@@ -415,9 +415,13 @@ impl BasicBlockEnvShape {
             | BaseTy::RawPtr(_, _)
             | BaseTy::Char
             | BaseTy::Never
-            | BaseTy::Closure(_, _)
-            | BaseTy::Dynamic(_, _)
-            | BaseTy::Coroutine(..) => bty.clone(),
+            | BaseTy::Closure(..)
+            | BaseTy::Dynamic(..)
+            | BaseTy::FnPtr(..)
+            | BaseTy::Coroutine(..) => {
+                debug_assert!(!scope.has_free_vars(bty));
+                bty.clone()
+            }
         }
     }
 

--- a/tests/tests/pos/fn_ptrs/fn_ptr00.rs
+++ b/tests/tests/pos/fn_ptrs/fn_ptr00.rs
@@ -1,0 +1,9 @@
+// Test that we at least support mentioning function pointers
+
+pub struct Struct {
+    f: fn(i32) -> i32,
+}
+
+pub fn test00(x: fn(i32) -> i32) {}
+
+pub fn test01(x: Vec<fn(i32) -> i32>) {}

